src/main/java/com/example/demo/services/impl/UserServiceImpl.java

```diff
@@ -1,49 +1,70 @@
+import com.example.demo.model.dto.User;
+import com.example.demo.model.dto.UserCreation;
+import com.example.demo.model.dto.UserUpdate;
+import com.example.demo.model.entities.UserEntity;
+import com.example.demo.repositories.UserRepository;
 import com.example.demo.services.UserService;
 import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
 
-import java.util.HashMap;
+import java.util.Date;
 import java.util.List;
-import java.util.Map;
+import java.util.Optional;
 
 @Service
 public class UserServiceImpl implements UserService {
 
-    private final Map<String, User> users = new HashMap<>();
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    private final UserEntityMapper userEntityMapper;
+    public UserServiceImpl(UserRepository userRepository, UserMapper userMapper, UserEntityMapper userEntityMapper) {
+        this.userRepository = userRepository;
+        this.userMapper = userMapper;
+        this.userEntityMapper = userEntityMapper;
+    }
+
 
     @Override
-    public Map<String, User> getUsers() {
+    public List<User> getUsers() {
+        List<UserEntity> userEntities = userRepository.findAll();
+        List<User> users = userMapper.map(userEntities);
         return users;
     }
 
     @Override
-    public List<User> getUsersByAge(final int min, final int max) {
-        return users.values().stream()
-                .filter(user -> user.getAge() >= min && user.getAge() <= max)
-                .toList();
+    public List<User> getUsersByAgeRange(final int min, final int max) {
+        List<User> users = userRepository.findByAgeBetween(min, max);
+        return users;
     }
 
     @Override
-    public User getUserByName(final String name) {
-        return users.get(name);
+    public User getUserById(final long id) {
+        return userMapper.map(userRepository.getById(id));
     }
 
     @Override
-    public User addUser(final User user) {
-        users.put(user.getName(), user);
-        return user;
+    public String addUser(UserCreation user) {
+        userRepository.save(userEntityMapper.map(user));
+        return "User has been added";
     }
 
-    @Override
-    public User changeAge(final User user) {
-        users.put(user.getName(), user);
-        return users.get(user.getName());
+    @Transactional
+    public String updateUser(final UserUpdate userUpdate) {
+        Date creationDate = userRepository.getById(userUpdate.getId()).getCreatedAt();
+        UserEntity updatedUserEntity = userEntityMapper.map(userUpdate);
+        updatedUserEntity.setCreatedAt(creationDate);
+        userRepository.save(updatedUserEntity);
+        return "User has been updated";
     }
 
     @Override
-    public void deleteUser(final String name) {
-        users.remove(name);
+    public String deleteUser(final long id) {
+        userRepository.delete(userRepository.getById(id));
```

Comment:
It would be more optimal to user `userRepository.deleteById()`. In that case you don't need to go twice to the DB